File: TI1.1 SOAP_JAVA_CONUNI_GR03/02 CLIMOV/app/src/main/java/ec/edu/espe/soap_conuni_clmov_gro3/arguello/FirstFragment.java
package ec.edu.espe.soap_conuni_clmov_gro3.arguello;

import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import ec.edu.espe.soap_conuni_clmov_gro3.R;
import ec.edu.espe.soap_conuni_clmov_gro3.arguello.servicios.ConUniServicio;
import ec.edu.espe.soap_conuni_clmov_gro3.databinding.FragmentFirstBinding;

public class FirstFragment extends Fragment {

    private FragmentFirstBinding binding;
    private ConUniServicio conversionService;
    private static final String TAG = "FirstFragment";
    
    private String[] conversionTypes = {"MASA", "TEMPERATURA", "LONGITUD"};
    private String[][] unitsByType = {
        {"Miligramo (mg)", "Gramo (g)", "Kilogramo (kg)", "Libra (lb)", "Onza (oz)", "Tonelada (t)"},
        {"Celsius (c)", "Fahrenheit (f)", "Kelvin (k)", "Rankine (r)"},
        {"Milímetro (mm)", "Centímetro (cm)", "Metro (m)", "Kilómetro (km)", "Pulgada (in)", "Pie (ft)"}
    };
    private String[][] unitCodesByType = {
        {"mg", "g", "kg", "lb", "oz", "t"},
        {"c", "f", "k", "r"},
        {"mm", "cm", "m", "km", "in", "ft"}
    };
    
    private int currentTypeIndex = 0;

    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState
    ) {
        binding = FragmentFirstBinding.inflate(inflater, container, false);
        conversionService = new ConUniServicio();
        return binding.getRoot();
    }

    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupConversionTypeSpinner();
        setupUnitSpinners();
        
        binding.convertButton.setOnClickListener(v -> performConversion());
    }
    
    private void setupConversionTypeSpinner() {
        ArrayAdapter<String> adapter = new ArrayAdapter<>(
            requireContext(),
            R.layout.spinner_item,
            conversionTypes
        );
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item);
        binding.conversionTypeSpinner.setAdapter(adapter);
        
        binding.conversionTypeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                currentTypeIndex = position;
                updateUnitSpinners();
                binding.resultCard.setVisibility(View.GONE);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }
    
    private void setupUnitSpinners() {
        updateUnitSpinners();
    }
    
    private void updateUnitSpinners() {
        String[] units = unitsByType[currentTypeIndex];
        
        ArrayAdapter<String> adapter = new ArrayAdapter<>(
            requireContext(),
            R.layout.spinner_item,
            units
        );
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item);
        
        binding.unitFromSpinner.setAdapter(adapter);
        binding.unitToSpinner.setAdapter(adapter);
        
        if (units.length > 1) {
            binding.unitToSpinner.setSelection(1);
        }
    }

    private void performConversion() {
        String valueStr = binding.valueInput.getText().toString().trim();
        
        if (valueStr.isEmpty()) {
            Toast.makeText(requireContext(), "Ingrese un valor", Toast.LENGTH_SHORT).show();
            return;
        }
        
        int fromIndex = binding.unitFromSpinner.getSelectedItemPosition();
        int toIndex = binding.unitToSpinner.getSelectedItemPosition();
        
        if (fromIndex == toIndex) {
            Toast.makeText(requireContext(), "Las unidades deben ser diferentes", Toast.LENGTH_SHORT).show();
            return;
        }

        binding.convertButton.setEnabled(false);
        binding.errorText.setVisibility(View.GONE);
        
        String fromUnit = unitCodesByType[currentTypeIndex][fromIndex];
        String toUnit = unitCodesByType[currentTypeIndex][toIndex];
        
        new ConversionTask().execute(valueStr, fromUnit, toUnit);
    }
    
    private String formatNumber(double value) {
        double absValue = Math.abs(value);
        
        if ((absValue > 0 && absValue < 0.001) || absValue >= 1_000_000) {
            return String.format("%.6e", value);
        } else if (absValue >= 1000) {
            return String.format("%,.3f", value);
        } else {
            return String.format("%.3f", value);
        }
    }
    
    private class ConversionTask extends AsyncTask<String, Void, ConversionResult> {
        
        @Override
        protected ConversionResult doInBackground(String... params) {
            ConversionResult result = new ConversionResult();
            
            try {
                double value = Double.parseDouble(params[0]);
                String fromUnit = params[1];
                String toUnit = params[2];
                
                result.value = value;
                result.fromUnit = unitsByType[currentTypeIndex][binding.unitFromSpinner.getSelectedItemPosition()];
                result.toUnit = unitsByType[currentTypeIndex][binding.unitToSpinner.getSelectedItemPosition()];
                result.resultValue = conversionService.conversion(value, fromUnit, toUnit);
                result.success = true;
                
            } catch (NumberFormatException e) {
                Log.e(TAG, "Error de formato de numero", e);
                result.error = "Valor inválido";
            } catch (Exception e) {
                Log.e(TAG, "Error en conversion", e);
                result.error = "Error al conectar con el servidor: " + e.getMessage();
            }
            
            return result;
        }
        
        @Override
        protected void onPostExecute(ConversionResult result) {
            binding.convertButton.setEnabled(true);
            
            if (!result.success) {
                binding.errorText.setText(result.error);
                binding.errorText.setVisibility(View.VISIBLE);
                Toast.makeText(requireContext(), result.error, Toast.LENGTH_SHORT).show();
                binding.resultCard.setVisibility(View.GONE);
                return;
            }
            
            String resultText = formatNumber(result.value) + " " + result.fromUnit + 
                              "\n=\n" + formatNumber(result.resultValue) + " " + result.toUnit;
            
            binding.resultText.setText(resultText);
            binding.resultCard.setVisibility(View.VISIBLE);
            binding.errorText.setVisibility(View.GONE);
            Toast.makeText(requireContext(), "Conversión exitosa", Toast.LENGTH_SHORT).show();
        }
    }
    
    private static class ConversionResult {
        boolean success = false;
        String error = "";
        double value;
        double resultValue;
        String fromUnit;
        String toUnit;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

}

File: TI1.3 REST_JAVA_CONUNI_GRP3/02 CLIMOV/app/src/main/java/ec/edu/espe/conuni_restfull_java/FirstFragment.kt
package ec.edu.espe.conuni_restfull_java

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import ec.edu.espe.conuni_restfull_java.data.api.RetrofitClient
import ec.edu.espe.conuni_restfull_java.data.model.ConUni
import ec.edu.espe.conuni_restfull_java.databinding.FragmentFirstBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class FirstFragment : Fragment() {

    private var _binding: FragmentFirstBinding? = null
    private val binding get() = _binding!!
    
    private val apiService = RetrofitClient.apiService
    private val TAG = "FirstFragment"
    
    private val conversionTypes = arrayOf("MASA", "TEMPERATURA", "LONGITUD")
    private val unitsByType = arrayOf(
        arrayOf("Miligramo (mg)", "Gramo (g)", "Kilogramo (kg)", "Libra (lb)", "Onza (oz)", "Tonelada (t)"),
        arrayOf("Celsius (°C)", "Fahrenheit (°F)", "Kelvin (K)", "Rankine (°R)"),
        arrayOf("Milímetro (mm)", "Centímetro (cm)", "Metro (m)", "Kilómetro (km)", "Pulgada (in)", "Pie (ft)")
    )
    private val unitCodesByType = arrayOf(
        arrayOf("mg", "g", "kg", "lb", "oz", "t"),
        arrayOf("c", "f", "k", "r"),
        arrayOf("mm", "cm", "m", "km", "in", "ft")
    )
    private val typeMapping = mapOf(
        0 to "masa",
        1 to "temperatura",
        2 to "longitud"
    )
    
    private var currentTypeIndex = 0

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentFirstBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        
        setupConversionTypeSpinner()
        setupUnitSpinners()
        
        binding.convertButton.setOnClickListener {
            performConversion()
        }
    }
    
    private fun setupConversionTypeSpinner() {
        val adapter = ArrayAdapter(
            requireContext(),
            R.layout.spinner_item,
            conversionTypes
        )
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item)
        binding.conversionTypeSpinner.adapter = adapter
        
        binding.conversionTypeSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                currentTypeIndex = position
                updateUnitSpinners()
                binding.resultCard.visibility = View.GONE
            }
            
            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }
    
    private fun setupUnitSpinners() {
        updateUnitSpinners()
    }
    
    private fun updateUnitSpinners() {
        val units = unitsByType[currentTypeIndex]
        
        val adapter = ArrayAdapter(
            requireContext(),
            R.layout.spinner_item,
            units
        )
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item)
        
        binding.unitFromSpinner.adapter = adapter
        binding.unitToSpinner.adapter = adapter
        
        if (units.size > 1) {
            binding.unitToSpinner.setSelection(1)
        }
    }
    
    private fun performConversion() {
        val valueStr = binding.valueInput.text.toString().trim()
        
        if (valueStr.isEmpty()) {
            Toast.makeText(requireContext(), "Ingrese un valor", Toast.LENGTH_SHORT).show()
            return
        }
        
        val value: Double
        try {
            value = valueStr.toDouble()
        } catch (e: NumberFormatException) {
            Toast.makeText(requireContext(), "Valor inválido", Toast.LENGTH_SHORT).show()
            return
        }
        
        val fromIndex = binding.unitFromSpinner.selectedItemPosition
        val toIndex = binding.unitToSpinner.selectedItemPosition
        
        if (fromIndex == toIndex) {
            Toast.makeText(requireContext(), "Las unidades deben ser diferentes", Toast.LENGTH_SHORT).show()
            return
        }
        
        binding.convertButton.isEnabled = false
        binding.errorText.visibility = View.GONE
        
        val fromUnit = unitCodesByType[currentTypeIndex][fromIndex]
        val toUnit = unitCodesByType[currentTypeIndex][toIndex]
        val type = typeMapping[currentTypeIndex] ?: "masa"
        
        lifecycleScope.launch {
            try {
                val result = withContext(Dispatchers.IO) {
                    val conUni = ConUni(type, value, fromUnit, toUnit)
                    val response = apiService.convertUnit(conUni)
                    if (response.isSuccessful && response.body() != null) {
                        response.body()!!
                    } else {
                        throw Exception("Error en la conversión: ${response.code()}")
                    }
                }
                
                val fromUnitText = unitsByType[currentTypeIndex][fromIndex]
                val toUnitText = unitsByType[currentTypeIndex][toIndex]
                
                val resultText = "${formatNumber(value)} $fromUnitText\n=\n${formatNumber(result)} $toUnitText"
                
                binding.resultText.text = resultText
                binding.resultCard.visibility = View.VISIBLE
                binding.errorText.visibility = View.GONE
                Toast.makeText(requireContext(), "Conversión exitosa", Toast.LENGTH_SHORT).show()
                
            } catch (e: Exception) {
                Log.e(TAG, "Error en conversión", e)
                binding.errorText.text = "Error al conectar con el servidor: ${e.message}"
                binding.errorText.visibility = View.VISIBLE
                Toast.makeText(requireContext(), "Error: ${e.message}", Toast.LENGTH_SHORT).show()
                binding.resultCard.visibility = View.GONE
            } finally {
                binding.convertButton.isEnabled = true
            }
        }
    }
    
    private fun formatNumber(value: Double): String {
        val absValue = Math.abs(value)
        
        return when {
            (absValue > 0 && absValue < 0.001) || absValue >= 1_000_000 -> {
                String.format("%.6e", value)
            }
            absValue >= 1000 -> {
                String.format("%,.3f", value)
            }
            else -> {
                String.format("%.3f", value)
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}